src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,343 @@
+            res.e_tag
+                .ok_or(anyhow!("upload_part response missing etag"))?,
+        );
+        self.state.current_part_number = next_part_number;
+        self.state.total_bytes_uploaded += num_of_bytes;
+        Ok(MultiPartProgress {
+            num_of_parts: self
+                .state
+                .current_part_number
+                .try_into()
+                .expect("i32 to u32"),
+            total_bytes_uploaded: self.state.total_bytes_uploaded,
+        })
+    }
+
+    /// Uploads a part for an ongoing multi-part upload.
+    /// Can only be called after calling `S3MultiPartUploader::start`
+    /// and can be called as many times required to upload each individual parts.
+    /// The total number of parts cannot exceed `MAX_PART_NUM` and
+    /// each part size should at least be `MIN_PART_SIZE` in size.
+    /// Returns the number of parts and number of bytes uploaded till now.
+    pub async fn upload_part(
+        &mut self,
+        data: Bytes,
+    ) -> Result<MultiPartProgress, S3MultiPartUploadError> {
+        let num_of_bytes: u64 = data.len().try_into().expect("usize to u64");
+        if num_of_bytes < MIN_PART_SIZE.as_u64() {
+            return Err(S3MultiPartUploadError::AtLeastMinPartSize);
+        }
+
+        self.upload_part_internal(data).await
+    }
+
+    /// Method to finish the multi part upload.
+    /// Takes optional `data` for a final part, which will
+    /// not be required to be at least `MIN_PART_SIZE`.
+    /// Returns the number of parts and number of bytes uploaded till now.
+    pub async fn finish(
+        mut self,
+        data: Option<Bytes>,
+    ) -> Result<MultiPartProgress, S3MultiPartUploadError> {
+        if let Some(data) = data {
+            let _ = self.upload_part_internal(data).await?;
+        }
+
+        let InProgress {
+            upload_id,
+            etags,
+            total_bytes_uploaded,
+            current_part_number,
+        } = self.state;
+
+        if current_part_number < MIN_PART_NUM {
+            return Err(S3MultiPartUploadError::AtLeastMinPartNumber);
+        }
+
+        let parts: Vec<CompletedPart> = etags
+            .iter()
+            .map(|(part_num, etag)| {
+                CompletedPart::builder()
+                    .e_tag(etag)
+                    .part_number(*part_num)
+                    .build()
+            })
+            .collect();
+        self.client
+            .complete_multipart_upload()
+            .bucket(&self.bucket)
+            .key(&self.path)
+            .upload_id(upload_id)
+            .multipart_upload(
+                CompletedMultipartUpload::builder()
+                    .set_parts(Some(parts))
+                    .build(),
+            )
+            .send()
+            .await?;
+        Ok(MultiPartProgress {
+            num_of_parts: current_part_number.try_into().expect("i32 to u32"),
+            total_bytes_uploaded,
+        })
+    }
+}
```

Comment:
Don't need to do this in this PR, but as mentioned live I think it'd be cool to explore having this utility also handle the chunking into multiple blobs, to stay under some max blob size that's configurable per `S3MultipartUploader`, and ultimately wired up to a user settable option in the `COPY TO S3` command.

Reply:
Ack. I had made this a sort of dumb utility because I was not sure how big usually batches of data in timely are (or maybe I am thinking about this wrong). Like are they big enough to be chunked into multiple parts or small enough that they should be buffered before uploading a part. 

I am updating this now to have a configurable max file size, configurable max part size as well, with defaults so that it can handle both cases, where `upload_part` sends small amount of data which can be buffered before uploading a part. And it can also send say a large amount of bytes (but not exceeding the max file size) which can be broken into multiple parts for the uploads.

Reply:
> Like are they big enough to be chunked into multiple parts or small enough that they should be buffered before uploading a part.

Both! Batches can be very large or very small. 

> I am updating this now to have a configurable max file size, configurable max part size as well, with defaults so that it can handle both cases, where `upload_part` sends small amount of data which can be buffered before uploading a part. And it can also send say a large amount of bytes (but not exceeding the max file size) which can be broken into multiple parts for the uploads.

It occurs to me that splitting files will need to be format aware. You wouldn't want to split a CSV in the middle of a line, for example. So I think I retract my earlier suggestion to try to handle a maximum file size here. We'll need to do that at the layer where we handle the CSV encoding.

Reply:
Yes, splitting files will need to be format aware, the smallest unit it can split by should be a row.

Here, the uploader is specific to one single file, so the maximum file size is more like an overall limit for the number of bytes which can be added to be uploaded.